From 00f8f358da42e01e31cea9013802188b2a1b7435 Mon Sep 17 00:00:00 2001
From: sjorge <sjorge@blackdot.be>
Date: Fri, 27 Sep 2019 11:19:34 +0200
Subject: [PATCH] OS-XXXX Do not destroy external delegated dataset

Do not recursively destroy dataset that are external to the zonepath
 instead return and remount them on the global zone.
---
 usr/src/lib/brand/jcommon/libhooks.ksh | 14 ++++++++++----
 1 file changed, 10 insertions(+), 4 deletions(-)

diff --git a/usr/src/lib/brand/jcommon/libhooks.ksh b/usr/src/lib/brand/jcommon/libhooks.ksh
index 346356519c..dabd0380d4 100644
--- a/usr/src/lib/brand/jcommon/libhooks.ksh
+++ b/usr/src/lib/brand/jcommon/libhooks.ksh
@@ -16,7 +16,7 @@
 #
 
 #
-# This file contains various hooks that are used by nore than a single
+# This file contains various hooks that are used by more than a single
 # brand. This file should be included by the brand-specific files.
 #
 
@@ -87,10 +87,16 @@ jattach_zone_final_setup()
 
 function juninstall_delegated_dataset
 {
-	# Now destroy any delegated datasets. Redirect to /dev/null in case they
-	# were already destroyed when we removed the zonepath dataset.
+	# Reset the zoned property for any delegated datasets. Also try and remount them
+	# in the global zone. Redirect to /dev/null in case any were destroyed when 
+	# we removed the zonepath dataset.
 	DD=`zonecfg -z $ZONENAME info dataset | nawk '{if ($1 == "name:") print $2}'`
 	for i in $DD; do
-		zfs destroy -rF $i >/dev/null 2>&1
+		zfs inherit zoned -r $i >/dev/null 2>&1
+		REMOUNT=`zfs list -o name,canmount,mounted,zoned -H -p -r $i | nawk '{if ($2 == "on" && $3 == "no" && $4 == "off") print $1 }'`
+		for j in $REMOUNT; do
+			zfs mount $j >/dev/null 2>&1
+		done
 	done
+
 }
-- 
2.21.0

